tensorflow_encrypted/tensor/odd_implicit.py

```diff
@@ -0,0 +1,126 @@
+        self.value = value
+        self.dtype = tf.int32
+        self.modulus = modulus
+
+    @staticmethod
+    def from_native(value: Union[np.ndarray, tf.Tensor], modulus: int, dtype=tf.int32) -> 'OddImplicitTensor':
+        assert isinstance(value, (np.ndarray, tf.Tensor)), type(value)
+        return OddImplicitTensor(value, modulus, dtype=dtype)
+
+    @staticmethod
+    def sample_uniform(shape: Union[Tuple[int, ...], tf.TensorShape], modulus: int, dtype=tf.int32) -> 'OddImplicitTensor':
+        return OddImplicitTensor(tf.random_uniform(shape=shape, dtype=dtype,
+                                                   minval=dtype.min, maxval=dtype.max), modulus, dtype)
+
+    @staticmethod
+    def sample_bounded(shape: List[int], bitlength: int) -> 'OddImplicitTensor':
+        raise NotImplementedError()
+
+    @staticmethod
+    def stack(x: List['OddImplicitTensor'], axis: int = 0) -> 'OddImplicitTensor':
+        assert all(isinstance(i, OddImplicitTensor) for i in x)
+        return OddImplicitTensor.from_native(tf.stack([v.value for v in x], axis=axis), x[0].modulus)
+
+    @staticmethod
+    def concat(x: List['OddImplicitTensor'], axis: int) -> 'OddImplicitTensor':
+        assert all(isinstance(i, OddImplicitTensor) for i in x)
+        return OddImplicitTensor.from_native(tf.concat([v.value for v in x], axis=axis), x[0].modulus)
+
+    def eval(self, sess: tf.Session, feed_dict: Dict[Any, Any]={},
+             tag: Optional[str]=None) -> 'OddImplicitTensor':
+        return OddImplicitTensor(run(sess, self.value, feed_dict=feed_dict, tag=tag), self.modulus)
+
+    def __getitem__(self, slice: Any) -> Union[tf.Tensor, np.ndarray]:
+        return self.value[slice]
+
+    def __repr__(self) -> str:
+        return 'OddImplicitTensor(shape={}, modulus={})'.format(self.shape, self.modulus)
+
+    @property
+    def shape(self) -> Union[Tuple[int, ...], tf.TensorShape]:
+        return self.value.shape
+
+    def __add__(self, other: Union['OddImplicitTensor', int]) -> 'OddImplicitTensor':
+        return self.add(other)
+
+    def __sub__(self, other: Union['OddImplicitTensor', int]) -> 'OddImplicitTensor':
+        return self.sub(other)
+
+    def __mul__(self, other: Union['OddImplicitTensor', int]) -> 'OddImplicitTensor':
+        return self.mul(other)
+
+    def __mod__(self, k: int) -> 'OddImplicitTensor':
+        return self.mod(k)
+
+    def add(self, other: Union['OddImplicitTensor', int]) -> 'OddImplicitTensor':
+        x, y = _lift(self, self.modulus, self.dtype), _lift(other, self.modulus, self.dtype)
+
+        z = x.value + y.value
+
+        def func(val):
+            return tf.cond(val >= self.modulus, lambda: val + 1, lambda: val)
+
+        ret = tf.map_fn(func, z)
+
+        return OddImplicitTensor(ret, self.modulus, dtype=self.dtype)
+
+    def sub(self, other: Union['OddImplicitTensor', int]) -> 'OddImplicitTensor':
```

Comment:
@mortendahl feel like this sub bit isn't right... we still need to compensate if we underflow?

Reply:
is this still a concern?

Reply:
I'm pretty sure its right, wouldn't mind a second pair of eyes though!